realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -1037,24 +1037,26 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
     return -1;
 }
 
-std::unique_ptr<Query> getHandoverQuery (jlong bgSharedGroupPtr, jlong queryPtr, bool advanceToLatestVersion)
+std::unique_ptr<Query> handoverQueryToWorker(jlong bgSharedGroupPtr, jlong queryPtr, bool advanceToLatestVersion)
 {
     SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, queryPtr);
     std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
 
-    SG(bgSharedGroupPtr)->end_read();
-
-    SharedGroup::VersionID currentVersion = SG(bgSharedGroupPtr)->get_version_of_current_transaction();
-    bool isDifferentVersions = (currentVersion != handoverQuery->version);
-    if (isDifferentVersions) {
-        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
-    } else {
-        SG(bgSharedGroupPtr)->begin_read();
-    }
-
     std::unique_ptr<Query> query = SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery));
-    if (advanceToLatestVersion && isDifferentVersions) {
+    if (advanceToLatestVersion) {
         LangBindHelper::advance_read(*SG(bgSharedGroupPtr));
+    } else {
+        try {
+            LangBindHelper::advance_read(*SG(bgSharedGroupPtr), handoverQuery->version);
```

Comment:
At this point in time, it may not be safe to access the handoverQuery. It is moved() at line 1045, so the access to version here is not well-defined.